[JSC] Optimze `String#at` in DFG/FTL

sosukesuzuki opened the pull request:
#### a509a5dd2a1c39679bfc2754492e55e5aa84660a
<pre>
[JSC] Optimze `String#at` in DFG/FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=217139">https://bugs.webkit.org/show_bug.cgi?id=217139</a>

Reviewed by Yusuke Suzuki.

This patch optimizes `String.prototype.at`. It reuses most of the optimizations made for
`String.prototype.charAt`, although there are some differences according to the specifications[1][2]:

- Returns undefined instead of an empty string when out of bounds
- Treats the `index` as `length + index` when the `index` is negative

                                            TipOfTree                  Patched

string-prototype-at-negative-index        2.0051+-0.0882     ^      1.1322+-0.0331        ^ definitely 1.7709x faster
string-prototype-at-out-of-bounds         1.4258+-0.0943     ^      1.1240+-0.0281        ^ definitely 1.2685x faster
string-prototype-at-positive-index        1.7500+-0.0490     ^      1.2082+-0.1981        ^ definitely 1.4484x faster

* JSTests/microbenchmarks/string-prototype-at-negative-index.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-out-of-bounds.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-positive-index.js: Added.
(test):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(at): Deleted.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
(JSC::DFG::ArrayMode::ArrayMode):
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByValOnString):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileStringCharAtImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileStringAt):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/287653@main">https://commits.webkit.org/287653@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ffd69c9a78ba5dba9553aa5606b162c4218820e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34101 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84962 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31423 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7750 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/62858 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/20664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83510 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52946 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73238 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43161 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/50255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27395 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29882 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73421 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27915 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86396 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79500 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7666 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71143 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7841 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69077 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70384 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17526 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14386 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13332 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101907 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7628 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24828 "Found 2 new JSC stress test failures: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-eager-jettison, wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7467 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/10987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9273 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -2517,15 +2519,24 @@ void SpeculativeJIT::compileGetByValOnString(Node* node, const ScopedLambda<std:
+
+        loadPtr(Address(baseReg, JSString::offsetOfValue()), scratchReg);
+        load32(Address(scratchReg, StringImpl::lengthMemoryOffset()), scratchReg);
+        neg32(propertyReg);
```
You cannot modify the register value which is still connected to some nodes.

Constellation replied:
Yeah, still this is changing `propertyReg`

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -2987,6 +2987,23 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
         }
 
+        case StringPrototypeAtIntrinsic: {
+            if (argumentCountIncludingThis < 2)
```
You can insert `is64Bit()` check and you can implement it for 64bit.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -2519,37 +2521,56 @@ void SpeculativeJIT::compileGetByValOnString(Node* node, const ScopedLambda<std:
     GPRReg scratchReg = resultRegs.payloadGPR();
 
+    GPRTemporary propertyTemp(this);
+    GPRReg propertyTempReg = propertyTemp.gpr();
```
Move this to L2512 etc. We should define these things in one place and materialize registers in on place

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -2519,37 +2524,53 @@ void SpeculativeJIT::compileGetByValOnString(Node* node, const ScopedLambda<std:
 
     if (node->op() == StringCharAt) {
         outOfBounds.link(this);
+        loadLinkableConstant(LinkableConstant(*this, jsEmptyString(vm())), resultRegs.payloadGPR());
```
32bit should put cell tag here, right?
These code originally exists,
```
if USE(JSVALUE32_64)
        if (format == DataFormatJS)
            move(TrustedImm32(JSValue::CellTag), resultRegs.tagGPR());
```

Why is it removed?

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -2519,37 +2524,53 @@ void SpeculativeJIT::compileGetByValOnString(Node* node, const ScopedLambda<std:
 #if USE(JSVALUE32_64)
-        if (format == DataFormatJS)
-            move(TrustedImm32(JSValue::CellTag), resultRegs.tagGPR());
+        move(TrustedImm32(JSValue::CellTag), resultRegs.tagGPR());
```
Why is it correct? (putting cell tag before `outOfBounds.link(this)`.

sosukesuzuki replied:
@Constellation  Ahh, Sorry. That is my mistake.  
Following the review at [https://github.com/WebKit/WebKit/pull/35997#discussion_r1845180154](https://github.com/WebKit/WebKit/pull/35997#discussion_r1845180154), I changed `StringAt` so that it is not optimized on 32-bit.  
As part of this change, I attempted to remove the 32 bit branching in `DFGSpeculativeJIT` for `StringAt`, but I mistakenly confused `StringAt` with `StringCharAt`.

I've fixed it.

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -10146,18 +10152,19 @@ IGNORE_CLANG_WARNINGS_END
 
         m_out.appendTo(is16Bit, bigCharacter);
 
-        LValue char16BitValue = m_out.load16ZeroExt32(baseIndexWithProvenValue(
```
Why not using `baseIndexWithProvenValue`? It looks like a regression.

sosukesuzuki commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -10146,9 +10152,11 @@ IGNORE_CLANG_WARNINGS_END
-        ValueFromBlock char8Bit = m_out.anchor(
-            m_out.load8ZeroExt32(baseIndexWithProvenValue(m_heaps.characters8, m_out.loadPtr(stringImpl, m_heaps.StringImpl_data), index,
-                m_graph.child(m_node, 1))));
+        LValue char8BitValue = m_out.load8ZeroExt32(m_out.baseIndex(
```
@Constellation I've addressed https://github.com/WebKit/WebKit/pull/35997/files#r1861247451, and I have questions:

- What is different between `baseIndex` and `baseIndexWithProvenValue`? What does "proven value" mean?
- ~I changed to use `baseIndexWithProvenValue` instead of `baseIndex` for `char8BitValue`, similar to the original code. However, this resulted in a incorrect value. Do you know why this might be happening?~ (this is my misunderstood)

Constellation replied:
`baseIndexWithProvenValue` is offering proven value too, and if the value is proven, then we attempt to optimize this based on the proven value. Thus, `baseIndexWithProvenValue` is more aggressive optimization than `baseIndex`. (So, converting `baseIndexWithProvenValue` to `baseIndex` is a regression).

sosukesuzuki replied:
@Constellation  Thank you.

In `StringAt`, unlike `StringCharAt`, if the index value of the original node is negative, it is not used as-is but calculated(string length + index) before use. So, using `baseIndexWithProvenValue` likely resulted in an incorrect value. To address this, I modified the implementation so that only when the node is `StringAt`, `m_out.baseIndex` is used instead of `baseIndexWithProvenValue`.

darinadler commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -2440,6 +2440,15 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+        else
+            setForNode(node, m_vm.stringStructure.get());
+        break;
+    }
```
Can write this instead:
```
if (node->arrayMode().isOutOfBounds())
```
Then won’t need the braces or local variable.

darinadler commented on Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
```diff
@@ -336,7 +336,8 @@ class BackwardsPropagationPhase : public Phase {
         case StringCharCodeAt:
-        case StringCodePointAt: {
+        case StringCodePointAt:
+        case StringAt: {
```
The others here are alphabetical so maybe put `StringAt` first instead of last.

darinadler commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -170,6 +170,7 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
         case StringCharAt:
         case StringCharCodeAt:
         case StringCodePointAt:
+        case StringAt:
```
Alphabetical order here too.

darinadler commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -2020,6 +2021,7 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
         return;
 
     case StringCharAt:
+    case StringAt:
```
Alphabetical order here too.

darinadler commented on Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
```diff
@@ -365,6 +365,7 @@ bool doesGC(Graph& graph, Node* node)
     case ResolveScopeForHoistingFuncDeclInEval:
     case Return:
     case StringCharAt:
+    case StringAt:
```
Alphabetical order here too.

darinadler commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -1077,10 +1077,14 @@ class FixupPhase : public Phase {
             break;
 
         case StringCharAt:
+        case StringAt:
```
Alphabetical order here too.

darinadler commented on Source/JavaScriptCore/dfg/DFGNode.h:
```diff
@@ -2544,6 +2544,7 @@ struct Node {
         case EnumeratorHasOwnProperty:
         case StringCharAt:
         case StringCharCodeAt:
+        case StringAt:
```
Alphabetical order here too.

darinadler commented on Source/JavaScriptCore/dfg/DFGNodeType.h:
```diff
@@ -353,6 +353,7 @@ namespace JSC { namespace DFG {
     macro(StringCodePointAt, NodeResultInt32) \
     macro(StringCharAt, NodeResultJS) \
     macro(StringFromCharCode, NodeResultJS | NodeMustGenerate) \
+    macro(StringAt, NodeResultJS) \
```
Alphabetical order here too.

darinadler commented on Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
```diff
@@ -640,6 +640,15 @@ class PredictionPropagationPhase : public Phase {
+
+        case StringAt: {
+            ArrayMode arrayMode = node->arrayMode();
+            if (arrayMode.isOutOfBounds())
```
Same thing again, don’t need the local variable and braces.

darinadler commented on Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
```diff
@@ -405,6 +405,7 @@ bool safeToExecute(AbstractStateType& state, Graph& graph, Node* node, bool igno
     case StringCharAt:
     case StringCharCodeAt:
     case StringCodePointAt:
+    case StringAt:
```
Alphabetical order here too.

darinadler commented on Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
```diff
@@ -161,6 +161,7 @@ inline CapabilityLevel canCompile(Node* node)
     case GetArgument:
     case InvalidationPoint:
     case StringCharAt:
+    case StringAt:
```
Alphabetical order here too.

sosukesuzuki committed:
[JSC] Optimze `String#at` in DFG/FTL
https://bugs.webkit.org/show_bug.cgi?id=217139

Reviewed by Yusuke Suzuki.

This patch optimizes `String.prototype.at`. It reuses most of the optimizations made for
`String.prototype.charAt`, although there are some differences according to the specifications[1][2]:

- Returns undefined instead of an empty string when out of bounds
- Treats the `index` as `length + index` when the `index` is negative

                                            TipOfTree                  Patched

string-prototype-at-negative-index        2.0051+-0.0882     ^      1.1322+-0.0331        ^ definitely 1.7709x faster
string-prototype-at-out-of-bounds         1.4258+-0.0943     ^      1.1240+-0.0281        ^ definitely 1.2685x faster
string-prototype-at-positive-index        1.7500+-0.0490     ^      1.2082+-0.1981        ^ definitely 1.4484x faster

* JSTests/microbenchmarks/string-prototype-at-negative-index.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-out-of-bounds.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-positive-index.js: Added.
(test):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(at): Deleted.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
(JSC::DFG::ArrayMode::ArrayMode):
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByValOnString):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileStringCharAtImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileStringAt):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/287653@main
